src/main/java/org/wildfly/security/auth/client/AuthenticationConfiguration.java

```diff
@@ -746,6 +746,55 @@ public AuthenticationConfiguration useCallbackHandler(CallbackHandler callbackHa
         return callbackHandler == null ? this : new AuthenticationConfiguration(this, SET_USER_CBH, callbackHandler, SET_USER_CB_KINDS, EnumSet.copyOf(callbackKinds));
     }
 
+    /**
+     * Create a new configuration which is the same as this configuration, but which uses the given supplier to make
+     * new callback handler to authenticate.
+     * <p>
+     * <em>Important notes:</em> It is important to ensure that each distinct client identity uses a distinct {@code CallbackHandler}
+     * instance in order to avoid mis-pooling of connections, identity crossovers, and other potentially serious problems.
+     * It is not recommended that a {@code CallbackHandler} implement {@code equals()} and {@code hashCode()}, however if it does,
+     * it is important to ensure that these methods consider equality based on an authenticating identity that does not
+     * change between instances.  In particular, a callback handler which requests user input on each usage is likely to cause
+     * a problem if the user name can change on each authentication request.
+     * <p>
+     * Because {@code CallbackHandler} instances are unique per identity, it is often useful for instances to cache
+     * identity information, credentials, and/or other authentication-related information in order to facilitate fast
+     * re-authentication.
+     *
+     * @param callbackHandlerSupplier the callback handler supplier to use
+     * @return the new configuration
+     */
+    public AuthenticationConfiguration useCallbackHandler(Supplier<CallbackHandler> callbackHandlerSupplier) {
+        if(callbackHandlerSupplier == null) return this;
+        CallbackHandler callbackHandler = callbackHandlerSupplier.get();
```

Comment:
The intent of ELY-929, in this particular case, was to use the supplier lazily, such that it would only be accessed when a protocol implementation actually acquires the configuration to use for a connection.  Evaluating the supplier directly in the setter adds little value and probably establishes a false expectation.

But note that even the intent of ELY-929 is possibly problematic/fragile, so this change shouldn't be entered in lightly either way.